File: python/tests/test_statement.py
import json

import pytest
from approvaltests import verify
from approval_utilities.utils import get_adjacent_file

from statement import statement


def test_example_statement():
    with open(get_adjacent_file("invoice.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("plays.json")) as f:
        plays = json.loads(f.read())
    verify(statement(invoice, plays))


def test_statement_with_new_play_types():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    with pytest.raises(ValueError) as exception_info:
        statement(invoice, plays)
    assert "unknown type" in str(exception_info.value)

File: cpp/tests/test_statement.cpp
#include "statement.h"

#include "json.hpp"
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest/doctest.h"
#define APPROVALS_DOCTEST
#include "ApprovalTests.hpp"

#include <iostream>

namespace
{
    std::string get_adjacent_file(const std::string& filename)
    {
        return ApprovalTests::ApprovalTestNamer().getDirectory() + filename;
    }

    nlohmann::json read_json_file(const std::string& filename)
    {
        std::ifstream invoice_stream(filename);
        if ( ! invoice_stream.is_open())
        {
            std::cout << "Error opening file:" << filename << '\n';
            std::cout << "Error: " << strerror(errno) << '\n';
        }
        nlohmann::json result;
        invoice_stream >> result;
        return result;
    }
}

TEST_CASE("test_example_statement")
{
    auto invoice = read_json_file(get_adjacent_file("invoice.json"));
    auto plays = read_json_file(get_adjacent_file("plays.json"));

    ApprovalTests::Approvals::verify(statement(invoice, plays));
}

TEST_CASE("test_statement_with_new_play_types")
{
    auto invoice = read_json_file(get_adjacent_file("invoice_new_plays.json"));
    auto plays = read_json_file(get_adjacent_file("new_plays.json"));
    CHECK_THROWS_AS(statement(invoice, plays), std::domain_error);
}
